libraries/Microsoft.Bot.Builder.Azure/AzureBlobStorage.cs

```diff
@@ -152,24 +155,40 @@ public AzureBlobStorage(CloudStorageAccount storageAccount, string containerName
             if (changes == null) throw new ArgumentNullException(nameof(changes));
 
             var blobContainer = await GetBlobContainer().ConfigureAwait(false);
+            var blobRequestOptions = new BlobRequestOptions();
+            var operationContext = new OperationContext();
 
             await Task.WhenAll(
                 changes.Select(async (keyValuePair) =>
                 {
                     var newValue = keyValuePair.Value;
                     var storeItem = newValue as IStoreItem;
                     // "*" eTag in IStoreItem converts to null condition for AccessCondition
-                    var calculatedETag = storeItem?.eTag == "*" ? null : storeItem?.eTag;
+                    var accessCondition = storeItem?.eTag == "*"
+                        ? AccessCondition.GenerateEmptyCondition()
+                        : AccessCondition.GenerateIfMatchCondition(storeItem?.eTag);
 
                     var blobName = GetBlobName(keyValuePair.Key);
                     var blobReference = blobContainer.GetBlockBlobReference(blobName);
-                    using (var blobStream = await blobReference.OpenWriteAsync(
-                        AccessCondition.GenerateIfMatchCondition(calculatedETag),
-                        new BlobRequestOptions(),
-                        new OperationContext()))
-                    using (var jsonWriter = new JsonTextWriter(new StreamWriter(blobStream)))
+
+                    try
+                    {
+                        using (var memoryStream = new MultiBufferMemoryStream(blobReference.ServiceClient.BufferManager))
+                        using (var streamWriter = new StreamWriter(memoryStream))
+                        {
+                            JsonSerializer.Serialize(streamWriter, newValue);
+                            streamWriter.Flush();
+                            memoryStream.Seek(0, SeekOrigin.Begin);
+                            await blobReference.UploadFromStreamAsync(memoryStream, accessCondition, blobRequestOptions, operationContext);
+                        }
+                    }
+                    catch (StorageException ex)
+                    when (ex.RequestInformation.HttpStatusCode == (int)HttpStatusCode.BadRequest
+                    && ex.RequestInformation.ErrorCode == BlobErrorCodeStrings.InvalidBlockList)
                     {
-                        JsonSerializer.Serialize(jsonWriter, newValue);
+                        throw new BotStorageException(
+                            $"An error ocurred while trying to write an object. The underlying '{BlobErrorCodeStrings.InvalidBlockList}' error is commonly caused due to concurrently uploading an object larger than 128MB in size.",

```

Comment:
I don't know if I'd hardcode this exact size in the message... maybe use `{blobReference.StreamWriteSizeInBytes}`?

Also, it's `4MB` by default, isn't it??? Did we choose to increase the default `StreamWriteSizeInBytes` somewhere that I'm not seeing or did I misunderstand the default size?

Reply:
We took this message in this [doc comment](https://github.com/Azure/azure-storage-net/blob/master/Lib/Common/Blob/BlobRequestOptions.cs#L41-L45):
```
        /// <summary>
        /// Indicates the maximum size of a blob, in bytes, that may be uploaded as a single blob,
        /// ranging from between 1 and 256 MB inclusive. The Default is 128 MB.
        /// </summary>
        private long? singleBlobUploadThresholdInBytes;
```
Also, check this doc https://docs.microsoft.com/es-es/rest/api/storageservices/put-blob#remarks

The 4MB limit is for each block uploaded once the SDK decided it has to be uploaded in blocks and not in a single PUT.